File: Java/leetcode75/src/main/java/org/phoenix/leetcode/challenges/Problem13_MaxNumberOfKSumPairs.java
package org.phoenix.leetcode.challenges;

import java.util.HashMap;
import java.util.Map;

public class Problem13_MaxNumberOfKSumPairs {
    public int maxOperations(int[] nums, int k) {
        int pairs = 0;
        Map<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        for (int key : map.keySet()) {
            pairs += Math.max(map.getOrDefault(key, 0), map.getOrDefault(key - k, 0));
        }
        return pairs / 2;
    }
}

File: JavaScript/leetcode75/src/main/Problem13_MaxNumberOfKSumPairs.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maxOperations = function (nums, k) {
  let pairs = 0;
  let map = new Map();
  for (let num of nums) {
    if (num < k) {
      map.set(num, (map.get(num) || 0) + 1);
    }
  }
  for (let key of map.keys()) {
    pairs += Math.min(map.get(key) || 0, map.get(k - key) || 0);
  }

  return Math.floor(pairs / 2);
};

module.exports = maxOperations;
